src/fem/op_function.hh

```diff
@@ -0,0 +1,1169 @@
+            for (uint i_c=0; i_c<spacedim; ++i_c)
+                for (uint i_dim=0; i_dim<dim; ++i_dim)
+                    for (uint i_dof=0; i_dof<this->n_dofs_; ++i_dof)
+                        for (uint i_p=0; i_p<n_points; ++i_p) {
+                            ref_vector_value(i_sd*dim+i_dim, 3*i_dof+i_c)(i_p) = ref_shape_grads[i_sd][i_p][i_dof](i_dim, i_c);
+                        }
+    }
+
+    void eval() override {}
+};
+
+/// Evaluates scalar values
+template<unsigned int dim, unsigned int spacedim = 3>
+class OpScalarShape : public Op::Side::Base<dim, spacedim> {
+public:
+    /// Constructor
+    OpScalarShape(PatchFEValues<spacedim> &pfev, std::shared_ptr<FiniteElement<dim>> fe)
+    : Op::Side::Base<dim, spacedim>(pfev, {1}, OpSizeType::pointOp, fe->n_dofs())
+    {
+        ASSERT_EQ(fe->fe_type(), FEType::FEScalar).error("Type of FiniteElement of scalar_shape must be FEScalar!\n");
+        this->input_ops_.push_back( pfev.template get< OpRefScalar<dim, spacedim>, dim >(fe) );
+    }
+
+    void eval() override {
+        PatchPointValues<spacedim> &ppv = this->ppv();
+        this->allocate_result(ppv.n_points_, ppv.patch_arena());
+
+        auto ref_vec = this->input_ops(0)->result_matrix();
+        auto result_vec = this->result_matrix();
+
+        uint n_dofs = this->n_dofs();
+        uint n_sides = ppv.n_elems_;         // number of sides on patch
+        uint n_patch_points = ppv.n_points_; // number of points on patch
+
+        for (uint i_dof=0; i_dof<n_dofs; ++i_dof) {
+            for (uint i_pt=0; i_pt<n_patch_points; ++i_pt) {
+                result_vec(i_dof)(i_pt) = ref_vec(ppv.int_table_(4)(i_pt), i_dof)(i_pt / n_sides);
+            }
+        }
+    }
+};
+
+/// Evaluates vector values (FEType == FEVector)
+template<unsigned int dim, unsigned int spacedim = 3>
+class OpVectorShape : public Op::Side::Base<dim, spacedim> {
+public:
+    /// Constructor
+	OpVectorShape(PatchFEValues<spacedim> &pfev, std::shared_ptr<FiniteElement<dim>> fe, PatchOp<spacedim> &dispatch_op)
+    : Op::Side::Base<dim, spacedim>(pfev, {spacedim}, OpSizeType::pointOp, fe->n_dofs()), dispatch_op_(dispatch_op)
+    {
+        this->input_ops_.push_back( pfev.template get< OpRefVector<dim, spacedim>, dim >(fe) );
+	}
+
+    void eval() override {
+        PatchPointValues<spacedim> &ppv = this->ppv();
+        dispatch_op_.allocate_result(ppv.n_points_, ppv.patch_arena());
+
+        auto ref_shape_vec = this->input_ops(0)->result_matrix();  // dim+1 x spacedim
+        auto result_vec = dispatch_op_.result_matrix();            // spacdim x 1
+
+        uint n_dofs = this->n_dofs();
+        uint n_sides = ppv.n_elems_;
+        uint n_patch_points = ppv.n_points_;
+
+        for (uint c=0; c<3*n_dofs; c++)
+        	result_vec(c) = ArenaVec<double>(n_patch_points, ppv.patch_arena());
+
+        for (uint i_dof=0; i_dof<n_dofs; ++i_dof) {
+            for (uint i_pt=0; i_pt<n_patch_points; ++i_pt)
```

Comment:
use spacedim

Reply:
Replaced in all places where using of spacedim is possible.